File: July-LeetCoding-Challenge/10-Flatten-a-Multilevel-Doubly-Linked-List/Flatten-a-Multilevel-Doubly-Linked-List.cpp
class Solution1 {
public:
    Node* flatten(Node* head, Node* tail = nullptr) {
        if(!head) return nullptr;
        Node* read = head, *pre;
        while(read){
            if(read->child){
                Node* tmp = read->next;
                Node* connect = flatten(read->child, tmp);
                read->child = nullptr;
                read->next = connect;
                read->next->prev = read;                
            }
            pre = read;
            read = read->next;
        }
        if(tail)
            pre->next = tail, tail->prev = pre; 
        return head;
    }
};

File: July-LeetCoding-Challenge/10-Flatten-a-Multilevel-Doubly-Linked-List/Flatten-a-Multilevel-Doubly-Linked-List.cs
public class Solution {
    public Node Flatten(Node head) {
        if(head == null) return null;
        var stack = new Stack<Node>();
        var c = head;
        
        while(c != null){
            if(c.child != null){
                if(c.next != null){
                    stack.Push(c.next);
                }
                c.next = c.child;
                c.next.prev = c;
                c.child = null;
            }else if(c.next == null && stack.Count > 0){
                c.next = stack.Pop();
                c.next.prev = c;
            }
            
            c = c.next;
        }
        
        return head;        
    }
}

/*
// Definition for a Node.
public class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
}
*/